test/extended/router/h2spec.go

```diff
@@ -0,0 +1,289 @@
+				g.By(fmt.Sprintf("[%s] Running conformance tests against host %q", t.routeType, t.hostname))
+
+				// sampleForFailures when true runs
+				// the conformance tests for N
+				// iterations to identify flaking
+				// tests. This should only be set to
+				// true in development (i.e., not CI).
+				//
+				// This should be run every time we
+				// consume any new version of haproxy
+				// be that a major, minor or a micro
+				// update to continuously validate the
+				// set of test case IDs that fail.
+				if sampleForFailures := false; sampleForFailures {
+					runConformanceTestsAndLogAggregateFailures(oc, t, 50)
+					continue
+				}
+
+				testSuites := runConformanceTests(oc, t)
+				failures := failingTests(testSuites)
+				failureCount := len(failures)
+
+				g.By("Analyzing results")
+				for _, f := range failures {
+					if _, exists := t.knownFailures[f.ID()]; exists {
+						failureCount -= 1
+						e2e.Logf("[%s] TestCase ID: %q is a known failure; ignoring", t.routeType, f.ID())
+					} else {
+						e2e.Logf("[%s] TestCase ID: %q (%q) ****FAILED****", t.routeType, f.ID(), f.TestCase.ClassName)
+					}
+				}
+				o.Expect(failureCount).Should(o.BeZero(), "expected zero failures")
+			}
+		})
+	})
+})
+
+func failingTests(testSuites []*h2spec.JUnitTestSuite) []h2specFailingTest {
+	var failures []h2specFailingTest
+
+	for _, ts := range testSuites {
+		for i := 0; i < ts.Tests; i++ {
+			if ts.TestCases[i].Error != nil {
+				failures = append(failures, h2specFailingTest{
+					TestNumber: i + 1,
+					TestCase:   ts.TestCases[i],
+				})
+			}
+		}
+	}
+
+	return failures
+}
+
+func runConformanceTests(oc *exutil.CLI, t h2specRouteTypeTest) []*h2spec.JUnitTestSuite {
+	podName := "h2spec"
+	e2e.ExpectNoError(oc.KubeFramework().WaitForPodRunning(podName))
+
+	outputFile, err := ioutil.TempFile("/tmp", "runConformanceTests")
+	o.Expect(err).NotTo(o.HaveOccurred())
+	defer os.Remove(outputFile.Name())
+
+	g.By("Running the h2spec CLI test")
+	h2specOutput, h2specErr := e2e.RunHostCmd(oc.KubeFramework().Namespace.Name, podName, t.h2specCommand(outputFile.Name()))
+	if h2specErr != nil {
+		// We don't assert on a failure here; h2spec will exit
+		// with non-zero if _any_ test in the suite fails. We
+		// later assert on failing to copy the results,
+		// failing to read the results, failing to decode the
+		// results because there are no testsuites in the
+		// decoded results. We always expect at least one
+		// testsuite in the results.
+		e2e.Logf("[%s] h2spec: error=\n%s", t.routeType, h2specErr)
+	}
+	if h2specErr != nil && h2specErr.Error() != h2specOutput {
+		// Also log any output if different.
+		e2e.Logf("[%s] h2spec: output=\n%s", t.routeType, h2specOutput)
+	}
+
+	g.By(fmt.Sprintf("[%s] Fetching %q from the container", t.routeType, outputFile.Name()))
+	rsyncErr := oc.Run("rsync").Args(podName+":"+outputFile.Name(), filepath.Dir(outputFile.Name()), "--strategy=tar").Execute()
```

Comment:
If you were having some stability issue with rsync, what about using `oc cp`?

Reply:
Thanks - will try that in sample mode.

Reply:
I think this is using rsync/tar under the hood because I see: 

```
STEP: [reencrypt] Copying h2spec:"/tmp/runConformanceTests513746697" "/tmp/runConformanceTests513746697"
Mar 18 13:50:10.115: INFO: Running 'oc --namespace=e2e-test-router-h2spec-6ds6p --kubeconfig=/tmp/configfile831321424 cp h2spec:/tmp/runConformanceTests513746697 /tmp/runConformanceTests513746697'
tar: Removing leading `/' from member names
```

I switched to `oc cp` anyway.